From 326501581d550d0397a1405425d1304ecec87e32 Mon Sep 17 00:00:00 2001
From: Benjamin Lin <benjamin-jw.lin@mediatek.com>
Date: Fri, 12 Jul 2024 15:44:03 +0800
Subject: [PATCH 82/89] mtk: mac80211: fix incorrect VIF assignment for IEEE
 802.11 fragments

In WDS mode, first fragment is obtained from ieee80211_local->fq, so its VIF is later changed from AP_VLAN to AP VIF.
On the other hand, subsequent fragments are obtained from txq_info->frags, so its VIF remains AP_VLAN.
Inconsistency in VIFs of fragments results in transmission failure.
Therefore, VIF assignment for non-first fragments is added.

Signed-off-by: Benjamin Lin <benjamin-jw.lin@mediatek.com>
---
 net/mac80211/tx.c | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/net/mac80211/tx.c b/net/mac80211/tx.c
index ebe3ae2..c6d3f2f 100644
--- a/net/mac80211/tx.c
+++ b/net/mac80211/tx.c
@@ -3850,8 +3850,11 @@ begin:
 	skb = __skb_dequeue(&txqi->frags);
 	if (unlikely(skb)) {
 		if (!(IEEE80211_SKB_CB(skb)->control.flags &
-				IEEE80211_TX_INTCFL_NEED_TXPROCESSING))
+				IEEE80211_TX_INTCFL_NEED_TXPROCESSING)) {
+			// TODO: report airtime of non-first fragments.
+			IEEE80211_SKB_CB(skb)->control.vif = vif;
 			goto out;
+		}
 		IEEE80211_SKB_CB(skb)->control.flags &=
 			~IEEE80211_TX_INTCFL_NEED_TXPROCESSING;
 	} else {
-- 
2.18.0

